### PR TITLE
feat(integer): add parallelized scalar rotate_left/right

### LINKED ISSUE #145: Add rotation (left/right) operation to integer digits
**What is the problem you want to solve and can not with the current version?**
At the moment, no rotate (left/right) operation is available for the shortint/integer digits.

**Describe the solution you'd like**
Adding to the shortint/integers API two methods:
- rotate_left
- rotate_right

**Describe alternatives you've considered**
Only manual implementation can offer an alternative at the moment.

### PR DESCRIPTION
**Pull request by tmontaigu**
Like shifts, rotates are implemented by combining
a rotation of the block and bivariate PBSs in case the rotation number `n` is not a multiple of the number of bits in a block.

Since the behaviour of rotations is to 'cycle' bits back the end/beginning of the 'bit slice' (i.e. no bits is ever lost like it can with shifts), the performance is always the same when (n % nb_bits_in block) != 0. However the implementation is simpler.

So assuming a machine where the number of threads
is >= to the ciphertext's number of block, the operation cost one bivariate PBS.

closes #145

### CODE REVIEW

File: tfhe/src/integer/server_key/radix_parallel/rotate.rs [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+            return;
+        }
+
+        let rotations = (n / num_bits_in_block) as usize;
+        let shift_within_block = n % num_bits_in_block;
+        let num_blocks = ct.blocks.len();
+
+        dbg!(rotations);
+
+        // rotate left as the blocks are from LSB to MSB
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/rotate.rs:623]: rotate right, there may be the same typo in the previous shift code

**Comment by tmontaigu** [comment]
> Looks generally good, I'm wondering if we maybe should have a special test case to be sure we handle the rotation part when bits "overflow/underflow" ?

We could but the random tests do not 'modularize' the random picked rotatation number, so that case il fairly certain to be tested i would believe

**Comment by IceTDrinker** [comment]
alright, it's just as seen before our statistical testing was not covering some edge cases enough, it can be a few hardcoded tests, or  even do as you did and lower the number of loops but with specialized test cases

**Comment by tmontaigu** [comment]
So what we can do is add a third case un the loop similar to the `// Force case where n is not multiple of block size`
and add `nb_bits` to the rotation number so we are guaranteed to have a overlow case

**Comment by IceTDrinker** [comment]
> So what we can do is add a third case in the loop similar to the `// Force case where n is not multiple of block size` and add `nb_bits` to the rotation number so we are guaranteed to have a overflow case

sounds like it should work well yes

**Comment by tmontaigu** [comment]
> > So what we can do is add a third case in the loop similar to the `// Force case where n is not multiple of block size` and add `nb_bits` to the rotation number so we are guaranteed to have a overflow case
> 
> sounds like it should work well yes

thinking about it, the thing i proposed is useless,
in rotations, bit alway circle around, if you rotate by 1, the then 1 bit circles back to the end/beginnin

**Comment by IceTDrinker** [comment]
well there is a chance the rotation is just a shift, I just want to be sure we cover the rotation case

File: tfhe/src/integer/server_key/radix_parallel/rotate.rs (hunk 1 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/rotate.rs]: I checked, shifts did not have this typo

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

**Review by IceTDrinker — APPROVED** [review]
Thanks a lot for the added test cases !
